system/src/ble_listening_mode_handler.cpp

```diff
@@ -41,110 +42,99 @@ BleListeningModeHandler::BleListeningModeHandler()
           prePpcp_(),
           preAdvertising_(false),
           preConnected_(false),
-          preAutoAdv_(BLE_AUTO_ADV_ALWAYS) {
+          preAutoAdv_(BLE_AUTO_ADV_ALWAYS),
+          userAdv_(false),
+          ctrlReqAdvData_(nullptr),
+          ctrlReqAdvDataLen_(0),
+          ctrlReqSrData_(nullptr),
+          ctrlReqSrDataLen_(0) {
 }
 
 BleListeningModeHandler::~BleListeningModeHandler() {
 }
 
-int BleListeningModeHandler::enter() {
-    // TODO: Would it be better to call the method of BLE Control Request Channel to isolate the logic?
-    // Cache the application specific advertising data, advertising parameters and connection parameters.
+int BleListeningModeHandler::constructControlRequestAdvData() {
+    ctrlReqAdvData_.reset(new(std::nothrow) uint8_t[BLE_MAX_ADV_DATA_LEN]);
+    CHECK_TRUE(ctrlReqAdvData_, SYSTEM_ERROR_NO_MEMORY);
+    ctrlReqSrData_.reset(new(std::nothrow) uint8_t[BLE_MAX_ADV_DATA_LEN]);
+    CHECK_TRUE(ctrlReqSrData_, SYSTEM_ERROR_NO_MEMORY);
+    ctrlReqAdvDataLen_ = 0;
+    ctrlReqSrDataLen_ = 0;
+
+    // AD Flag
+    ctrlReqAdvData_[ctrlReqAdvDataLen_++] = 0x02;
+    ctrlReqAdvData_[ctrlReqAdvDataLen_++] = BLE_SIG_AD_TYPE_FLAGS;
+    ctrlReqAdvData_[ctrlReqAdvDataLen_++] = BLE_SIG_ADV_FLAGS_LE_ONLY_GENERAL_DISC_MODE;
+
+    char devName[32] = {};
+    CHECK(get_device_name(devName, sizeof(devName)));
+
+    // Complete local name
+    ctrlReqAdvData_[ctrlReqAdvDataLen_++] = strlen(devName) + 1;
+    ctrlReqAdvData_[ctrlReqAdvDataLen_++] = BLE_SIG_AD_TYPE_COMPLETE_LOCAL_NAME;
+    memcpy(&ctrlReqAdvData_[ctrlReqAdvDataLen_], devName, strlen(devName));
+    ctrlReqAdvDataLen_ += strlen(devName);
+
+    uint16_t platformID = PLATFORM_ID;
+    uint16_t companyID = PARTICLE_COMPANY_ID;
+
+    // Manufacturing specific data
+    ctrlReqAdvData_[ctrlReqAdvDataLen_++] = sizeof(platformID) + sizeof(companyID) + 1;
+    ctrlReqAdvData_[ctrlReqAdvDataLen_++] = BLE_SIG_AD_TYPE_MANUFACTURER_SPECIFIC_DATA;
+    memcpy(&ctrlReqAdvData_[ctrlReqAdvDataLen_], (uint8_t*)&companyID, 2);
+    ctrlReqAdvDataLen_ += sizeof(companyID);
+    memcpy(&ctrlReqAdvData_[ctrlReqAdvDataLen_], (uint8_t*)&platformID, sizeof(platformID));
+    ctrlReqAdvDataLen_ += sizeof(platformID);
+
+    // Particle Control Request Service 128-bits UUID
+    ctrlReqSrData_[ctrlReqSrDataLen_++] = sizeof(BLE_CTRL_REQ_SVC_UUID) + 1;
+    ctrlReqSrData_[ctrlReqSrDataLen_++] = BLE_SIG_AD_TYPE_128BIT_SERVICE_UUID_COMPLETE;
+    memcpy(&ctrlReqSrData_[ctrlReqSrDataLen_], BLE_CTRL_REQ_SVC_UUID, sizeof(BLE_CTRL_REQ_SVC_UUID));
+    ctrlReqSrDataLen_ += sizeof(BLE_CTRL_REQ_SVC_UUID);
+    return SYSTEM_ERROR_NONE;
+}
+
+int BleListeningModeHandler::cacheUserConfigurations() {
+    LOG_DEBUG(TRACE, "Cache user's BLE configurations.");
+
+    // Advertising data set by user application
```

Comment:
We should acquire the BLE HAL lock for the whole duration of this function, in order to get a snapshot of the application BLE configuration without giving it a chance to change it in-between.

Also, how do we guard the application from changing the configuration while its running concurrently with the listening mode?

Reply:
How about introduce HAL API `hal_ble_notify_enter_listening_mode()` and `hal_ble_notify_exit_listening_mode()`? And if so, these two APIs will not be exported as part of dynalib.

Reply:
What would they do?

Reply:
They will set/clear a flag in BLE HAL to indicate whether the advertising data, advertising parameters, connection parameters and other configurations that might be changed during the mobile setup. Besides, we can also make use of this flag to handle some conflicts between the Listening mode and user application, such like `BLE.off()` being called in user application will then remain the peripheral connection if device is in the Listening mode.

Reply:
And I think with that HAL APIs implemented, it's unnecessary to acquire/release the BLE HAL lock in this function, since the configurations are guarded these APIs as well. No?

Reply:
As long as there is some mechanism in place that makes the current application configuration non-modifiable, yes.

Let's also make sure that we don't forget to call `hal_ble_notify_exit_listening_mode()` in case something goes wrong (e.g. memory allocation fails etc).